web/packages/shared/utils/highbar.test.ts

```diff
@@ -221,3 +226,89 @@ describe('arrayObjectIsEqual correctly compares', () => {
     expect(arrayObjectIsEqual(a, b)).toBe(true);
   });
 });
+
+describe('freezeDeep', () => {
+  it('freezes an empty object', () => {
+    const obj: { foo?: any } = freezeDeep({});
+    expect(() => (obj.foo = 'bar')).toThrow();
+  });
+  it('freezes a flat object', () => {
+    const obj = freezeDeep({ foo: 'bar' });
+    expect(() => (obj.foo = '123')).toThrow();
+  });
+  it('freezes a deep object', () => {
+    const obj = freezeDeep({ foo: { bar: 'baz' } });
+    expect(() => (obj.foo = { bar: '123' })).toThrow();
+    expect(() => (obj.foo.bar = '123')).toThrow();
+  });
+  it('freezes arrays', () => {
+    const obj = freezeDeep({ foo: ['bar'] });
+    expect(() => obj.foo.push('123')).toThrow();
+  });
+  it('returns the argument', () => {
+    const obj = {};
+    expect(freezeDeep(obj)).toBe(obj);
+  });
+});
+
+describe('equalsDeep', () => {
+  it('compares primitive values', () => {
+    expect(equalsDeep(1, 2)).toBe(false);
+    expect(equalsDeep(2, 2)).toBe(true);
+  });
+  it('compares simple objects', () => {
+    expect(equalsDeep({ a: 'b', c: 4 }, { c: 5, a: 'b' })).toBe(false);
+    expect(equalsDeep({ a: 'b', c: 4 }, { c: 4, a: 'b' })).toBe(true);
+  });
+  it('compares complex objects', () => {
+    expect(
+      equalsDeep(
+        { a: 'b', c: 1, d: [2, { e: 'f' }] },
+        { a: 'b', c: 5, d: [2, { e: 'f' }] }
+      )
+    ).toBe(false);
+    expect(
+      equalsDeep(
+        { a: 'b', c: 1, d: [2, { e: 'f' }] },
+        { a: 'b', c: 1, d: [5, { e: 'f' }] }
+      )
+    ).toBe(false);
+    expect(
+      equalsDeep(
+        { a: 'b', c: 1, d: [2, { e: 'f' }] },
+        { a: 'b', c: 1, d: [2, { e: 'f' }, 3] }
+      )
+    ).toBe(false);
+    expect(
+      equalsDeep(
+        { a: 'b', c: 1, d: [2, { e: 'f' }] },
+        { a: 'b', c: 1, d: [2, { e: 'z' }] }
+      )
+    ).toBe(false);
+    expect(
+      equalsDeep(
+        { a: 'b', c: 1, d: [2, { e: 'f' }] },
+        { a: 'b', c: 1, d: [2, { e: 'f' }], g: 'h' }
+      )
+    ).toBe(false);
+    expect(
+      equalsDeep(
+        { a: 'b', c: 1, d: [{ e: 'f' }, 2] },
+        { a: 'b', c: 1, d: [2, { e: 'f' }] }
+      )
+    ).toBe(false);
+    expect(
+      equalsDeep(
+        { a: 'b', c: 1, d: [2, { e: 'f' }] },
+        { a: 'b', z: 1, d: [2, { e: 'f' }] }
+      )
+    ).toBe(false);
+
+    expect(
+      equalsDeep(
+        { a: 'b', c: 1, d: [2, { e: 'f' }] },
+        { a: 'b', c: 1, d: [2, { e: 'f' }] }
+      )
+    ).toBe(true);
+  });
+});
```

Comment:
Can I ask for a test that tests deep equality and freezing on a proper role sample? The tests above looks good but its hard to judge how it fits to the actual purpose.

Reply:
I added an example with role options, since this is what it's gonna be used for.